feat: h2 DB 연결 및 테이블 생성, 리펙토링, BoardRepositoryTest 생성

jyeong2525 opened the pull request:
h2 DB 연결 및 BoardEntity 테이블 생성, 코드 정리, MVC패턴으로 진행되도록 리펙토링 진행함

Commit:
h2 DB 연결 및 테이블 생성, 리펙토링, BoardRepositoryTest 생성

h2 DB 연결 및 BoardEntity 테이블 생성, 코드 정리, MVC패턴으로 진행되도록 리펙토링 진행함

Hoyaspark commented on src/main/java/com/board/controller/BoardController.java:
```diff
@@ -16,8 +16,19 @@ public class BoardController {
+
     @GetMapping(path = "/board")
-    public ResponseEntity<List<Board>> getBoardList() {
+    public ResponseEntity<List<Board>> getBoards() {
```
[SUG] List<Board> 도 DTO 로 만들어서 보내주자

jyeong2525 replied:
완료

Hoyaspark commented on src/main/java/com/board/controller/BoardController.java:
```diff
@@ -37,30 +48,15 @@ public ResponseEntity<GetBoardResponseDto> getBoard(@PathVariable Long boardId)
         return ResponseEntity.ok().build();
     }
 
     @PatchMapping("/board/{boardId}")
```
[SUG] Patch 랑 Put 의 차이점 찾아서 공부

jyeong2525 replied:
PUT은 리소스의 전체를 새로운 데이터로 교체
PATCH는 리소스의 일부만을 수정

Hoyaspark commented on src/main/java/com/board/controller/BoardController.java:
```diff
@@ -37,30 +48,15 @@ public ResponseEntity<GetBoardResponseDto> getBoard(@PathVariable Long boardId)
-                boardId.getContent(),
-                boardId.getUserId());
+    public ResponseEntity<UpdateBoardRequestDto> updateBoard(@PathVariable Long boardId, @RequestBody UpdateBoardRequestDto updateBoardRequestDto) {
+        boardService.updateBoard(boardId, updateBoardRequestDto);
```
오 잘했네 boardId 받아서 boardId 해당하는 row 찾고 거기에 업데이트 시켜주기 굿

Hoyaspark commented on src/main/java/com/board/repository/BoardEntity.java:
```diff
@@ -1,13 +1,35 @@
-    private final String content;
-    private final String userId;
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
```
1. GenerationType.IDENTITY 와 그 외 여러개 어떨때 쓰는지 알아보기

2. GenerationType.IDENTITY 로 햇을떄 DB 에 실제로 어떻게 ID 를 순서대로 넣는지 알아보기

(힌트: JPA 는 DB에 INSERT 하기 전, 먼저 AUTO INCREMENT ID 값을 가져옴.)

jyeong2525 replied:
정리 완료

Hoyaspark commented on src/main/java/com/board/repository/BoardRepository.java:
```diff
@@ -11,24 +13,25 @@
 
     private Map<Long, BoardEntity> boardStore = new HashMap<>();
-    private Long id = 1L;
+    private Long id = 0L;
```
이건 더이상 필요없지 않을까? DB 에서 auto-incrementID 해서 알아서 id 는 집어넣어줄텐데

jyeong2525 replied:
삭제 완료

Hoyaspark commented on src/main/java/com/board/repository/BoardRepository.java:
```diff
@@ -11,24 +13,25 @@
     public void createBoard(Board board) {
-
-        BoardEntity boardEntity = new BoardEntity(id++, board.getTitle(), board.getContent(), board.getUserId());
+        BoardEntity boardEntity = new BoardEntity(++id, board.getTitle(), board.getContent(), board.getUserId());
```
++id 없애자. id 는 null 로 주면 됨

jyeong2525 replied:
변경 완료

Hoyaspark commented on src/test/java/com/board/repository/BoardRepositoryTest.java:
```diff
@@ -0,0 +1,75 @@
+    @Test
+    public void save() throws Exception{
+        //given
+        Board board = new Board(1L, "test", "test1", "userA");
```
id 를 명시해서 넣어주면 안되지. id 는 db 에서 알아서 결정해줄텐데. 그게 GenerationType.IDENTITY야

jyeong2525 replied:
완료

Hoyaspark reviewed (commented):
다시 한번 리뷰 보고 공부하기

Commit:
refactor : 피드백 사항 수정

getBoards 도 DTO로 보내도록 수정
GenerationType.IDENTITY를 사용해 id 부분 삭제

Commit:
feat: JPA 기능 추가

기존 map에서 JpaRepository로  JPA 기능 추가
DB는 mySQL 사용 관련 코드 추가
mydb 초기화 및 권한 생성
PostMan을 통해 테스트 완료

Commit:
test: BoardRepositoryTest 삭제

사용하지 않아서 삭제 처리

Hoyaspark commented on build.gradle:
```diff
@@ -29,7 +29,7 @@ dependencies {
 
     implementation 'org.springframework.boot:spring-boot-starter-data-jpa'
-    runtimeOnly("com.h2database:h2")
+    implementation 'mysql:mysql-connector-java:8.0.28'
```
[Q] 버전 명시한 이유가 뭐임? 그리고 mysql-connector 는 runtimeOnly 로 해도 될걸

Hoyaspark replied:
mysql:mysql-connector-java 로 했을때 안됐었다면, mysql:mysql-connector-j 로 해봐

이름 바뀜 버전 올라가면서

https://github.com/mysql/mysql-connector-j

jyeong2525 replied:
`implementaion`

- 컴파일과 런타임에 모두 필요한 종속성을 지정

`runtimeOnly`

- 런타임에만 필요한 종속성을 지정

수정 완료

Hoyaspark commented on src/main/java/com/board/board.sql:
```diff
@@ -0,0 +1,3 @@
+create DATABASE mydb;
+GRANT ALL PRIVILEGES ON mydb.* TO 'root'@'localhost';
```
[Q] 이건 무슨 용도??

jyeong2525 replied:
초기화와 권한 부여의 의미로 나오길래 썼습니당..

Hoyaspark commented on src/main/java/com/board/controller/JpaBoardController.java:
```diff
@@ -0,0 +1,51 @@
+    }
+
+    @GetMapping
+    public ResponseEntity<List<ListBoardResponseDto>> getBoards() {
```
ListBoardReponseDto 이름 자체에 List 를 내재하고 있는데, 이 객체를 List<ListBoardResponseDto> 로 하지 말고, ListBoardReponseDto 자체에서 List 를 멤버로 가지게 하면 더 코드가 깔끔할듯

일급객체 에 대해서 검색해봐

jyeong2525 replied:
일급객체(First Class Object)
- 다른 객체들에 일반적으로 적용 가능한 연산을 모두 지원하는 객체

Controller 수정 완료
[Q] Service 부분도 수정해야하는지?

Hoyaspark commented on src/main/java/com/board/repository/BoardEntity.java:
```diff
@@ -5,6 +5,7 @@
 
 @Data
```
@Data 로 열어주는거 안좋은 것 같은데.. 이거 엔티티 작성할떄 어떻게 해야하나 찾아봐

jyeong2525 replied:
JPA 엔티티에서는 여러가지 이유 (순환참조, @Setter 등)으로 사용하지 않는걸 권장한다네 
필요한 부분에 @Getter, @Setter 생성 완료

Hoyaspark commented on src/main/java/com/board/service/JpaBoardService.java:
```diff
@@ -0,0 +1,61 @@
+    }
+
+    public GetBoardResponseDto getBoard(Long boardId) {
+        BoardEntity boardEntity = jpaBoardRepository.findById(boardId)
```
ㅋㅋㅋㅋㅋ ㅅㅂ 잘했는데? ㅋㅋㅋ

Hoyaspark commented on src/main/resources/application.properties:
```diff
@@ -1,7 +1,9 @@
+spring.datasource.url=jdbc:mysql://localhost:3306/mydb
+spring.datasource.username=root
+
+spring.jpa.hibernate.ddl-auto=update
```
[Q] 이거 무슨뜻인지 아십니까?

jyeong2525 replied:
spring.jpa.hibernate.ddl-auto
: Hibernate의 DDL 자동 처리 전략
update
: 엔티티 구조에 맞춰 업데이트

Hoyaspark commented on src/main/resources/application.properties:
```diff
@@ -1,7 +1,9 @@
\ No newline at end of file
+
+spring.datasource.url=jdbc:mysql://localhost:3306/mydb
+spring.datasource.username=root
```
password 는 작성 안해도 됨?

jyeong2525 replied:
password를 지정 안해서 작성 안한거긴해 필요하면 지정하고 작성해야지 
일단 보류

Commit:
refactor : 리뷰사항 리펙토링, 일급 컬렉션 생성

BoardEntity @Data 삭제, @Getter 생성, @Setter 생성
ListBoardResponseDto에 @Getter로 변환

일급컬렉션 사용을 위해 ListBoardResponse 객체 생성, 기존 Json과 같도록 @JsonVaule 추가
JpaBoardController에 getBoards 코드 단순화

Commit:
refactor : ListBoardResponse 수정, board.sql 삭제

일급컬렉션 활용한 ListBoardResponse 수정
초기화, 권한 설정이 따로 필요없어 board.sql 삭제
